### PR TITLE
Fix upstream response decoder to close stream on trailers frame

### PR DESCRIPTION
**Pull request by ohadvano**
When using HTTP/2 upstream with TCP tunneling feature, the only place the reading is done (i.e using "doneReading()" callback) is in cases the last data frame from the upstream also contains the NGHTTP2_FLAG_END_STREAM flag. This way the decodeData callback will have "end_stream" to true.

In certain streaming cases, the downstream sends delayed FIN packet and the server will not respond with a data frame, rather than a headers frame which will be the trailers data and an END_STREAM flag. This causes "[doneReading()](https://github.com/envoyproxy/envoy/blob/main/source/common/tcp_proxy/upstream.h#:~:text=void%20decodeData(Buffer%3A%3AInstance%26%20data%2C%20bool%20end_stream)%20override)" callback to be skipped and the encoder would not be reset, which causes the filter chain not to drain.

### CODE REVIEW

**Comment by ohadvano** [comment]
@alyssawilk I am trying to avoid re-writing unit tests code in case the source code would change. Can you please confirm that the source changes are fine? I'll add the unit tests in another commit afterwards.

File: changelogs/current.yaml [hunk 1]
```diff
@@ -125,6 +125,9 @@ bug_fixes:
 - area: health_checker
   change: |
     prevent writing pending data for health checkers by introducing ``ConnectionCloseType::Abort`` to avoid cascading handshake overhead from health checker's requests on timeout. This fix is related to `issue \#23718 <https://github.com/envoyproxy/envoy/issues/23718>`_.
+- area: tcp_proxy
+  change: |
```
Comment (alyssawilk) [inline changelogs/current.yaml:129]: Let's change that first sentence to

When tunneling HTTP over TCP, mark the upstream connection as done reading when upstream trailers are read.

**Review by alyssawilk — COMMENTED** [review]
Code looks good modulo lack of tests.

**Comment by ohadvano** [comment]
For some reason, I had issues with the ADO build for this branch, even after many retries. Opened identical PR with same code: https://github.com/envoyproxy/envoy/pull/24800

**Pull request closed without merging**